### DISCUSSION TITLE
ERROR: find a bonded pair that is not on the same processor, something should not happen

### DISCUSSION THREAD

**Discussion by Constant206462** [deepmodeling/deepmd-kit#2319] [Q&A] [answered]
Hi! I'm using deepmd-kit 2.1.5. The error occurred on `run` command after applying `fix npt` while `nve` and `temp/berendsen` fixes lammps run just fine. 

> ERROR: find a bonded pair that is not on the same processor, something should not happen (/home/conda/feedstock_root/build_artifacts/libdeepmd_1663923207577/work/source/lmp/fix_dplr.cpp:211)
Last command: run		2000000

What i tried to do: change number of cpus, change atom groups in `fix npt`. Neither of those helped. Also `read_restart` from previous successful run don't fix the problem.
Inputs, logs and other files i use are [here](https://github.com/deepmodeling/deepmd-kit/files/10735458/proj.zip).
npt.in is just the last part of long_sim.in that reads nve restart files, error message was printed in both runs (slurm-11984 corresponds to long_sim, slurm-11988 - to npt). Hopefully it's not confusing.
Honestly, i have no idea how to get rid of this issue. Please help!

And i have another question. To which group of atoms `fix nve` and `fix temp/berendsen` should be applied? Is it `all` or only `real_atom`?

**Comment by wanghan-iapcm** [comment]
Please make sure that in the intial configuration, the positions of the WCs are the same as the real atoms which they are associated to.  
An example configuration can be found [here](https://docs.deepmodeling.com/projects/deepmd/en/master/model/dplr.html#molecular-dynamics-simulation-with-dplr), you may see that the position of atom 1 and 385 share the same position, 2 and 386 share the same and so on.

**Reply by Constant206462** [reply]
Many thanks for the answer. In the initial configuration (h2o.data file) WCs are placed right at corresponded real atoms. I noticed that after minimization run WCs happened to be displaced from O atoms. Then i removed whole `minimize` section, but at the second timestep in the trajectory file during `nve` + `temp/berendsen` i can see that WCs and O aren't share the same positions anymore even though displacement isn't large.

**Comment by ChunyiZh** [comment, marked as answer]
Please keep the positions of the WCs the same as the corresponding real atom every time you **restart** the simulation. 

During the simulation, the WCs positions and the real atom positions are different, which is correct. So, when you "write_restart", the WCs positions and the real atom positions are also different in your restart file. But this will cause a problem when you restart your simulation from your restart file. To make the WC and its corresponding real atom be allocated to the same processor, we need to change the WCs positions manually in the restart file. So, instead of "write_restart", I use "write_data out.lmp nocoeff". Then, I change the position of the WCs to make them equal to the real atom to enable a successful restart. Afterward, I use "read_data  xx" to restart my simulation.

Here is my code to change the WCs positions for a 64 water molecule system. You may need to write your own code according to your system.
[conf.txt](https://github.com/deepmodeling/deepmd-kit/files/10758833/conf.txt)

**Reply by Constant206462** [reply]
Thank you for your answer. I will keep in mind that. 
The main problem is that even if i don't restart the simulation (everything happens within one lammps script), the error would appear in `fix npt` part if there was another `run` with some ensemble (i tried either `nve` or `nvt`). But if i just run npt ensemble, everything is working, except system isn't relaxed. That's OK for test task with water but it might become a huge problem later.

**Reply by ChunyiZh** [reply]
You can try to use several Lammps scripts to do different ensembles, and use "write_data" and "read_data" to restart one ensemble from the end of the other ensemble.

**Reply by Constant206462** [reply]
Yeah, sounds like it's the way to go. Thanks again!
